lib/metasploit/framework/compiler/mingw.rb

```diff
@@ -43,90 +47,159 @@ def build_cmd(src)
-            cmd << self.compile_options
+          if compile_options
+            cmd << compile_options
           else
             link_options << '--image-base=0x0,'
             cmd << '-nostdlib '
           end
 
           link_options << '--no-seh'
-          link_options << ',-s' if self.strip_syms
-          link_options << ",-T#{self.link_script}" if self.link_script
+          link_options << ',-s' if strip_syms
+          link_options << ",-T #{link_script}" if link_script
 
           cmd << link_options
-
           cmd
         end
 
-        def cleanup_files
-          src_file = "#{self.file_name}.c"
-          exe_file = "#{self.file_name}.exe"
+        def compile_cpp(src, *additional_files)
+          within_temp(additional_files) do |dir, fnames|
+            file = Tempfile.create(['', '.cpp'], dir)
+            File.write(file, src)
+            if keep_src
+              dest = File.join(File.dirname(outfile), File.basename(file.path))
+              FileUtils.cp(file, dest)
+            end
+            fnames << file.path
+            cmd = build_cpp_files_cmd(fnames)
+            return exec(cmd), dest
+          end
+        end
+
+        def compile_cpp_files(files)
+          within_temp(files) do |_dir, fnames|
+            cmd = build_cpp_files_cmd(fnames)
+            exec(cmd)
+          end
+        end
+
+        def compile_cpp_file(file)
+          compile_cpp_files([file])
+        end
+
+        def build_cpp_files_cmd(file_array)
+          cmd = [ mingw_bin ]
+          cmd << file_array
+          cmd << '-I'
+          cmd << INCLUDE_DIR
+          cmd << '-o'
+          cmd << outfile
+          cmd << '-shared' if outfile.ends_with?('.dll')
+          cmd << opt_lvl if OPTIMIZATION_FLAGS.include?(opt_lvl)
+          cmd << compile_options
+
+          if link_options
+            link_opts = ['-Wl']
+            link_opts << '-s' if strip_syms
+            link_opts << link_options
+            cmd << link_opts.join(',')
+          end
+
+          if link_script
+            cmd << '-T'
+            cmd << link_script
+          end
+          cmd.flatten
+        end
+
+        def exec(cmd)
+          print("#{cmd.flatten.join(' ')}\n") if show_compile_cmd
+          stdout_err, status = Open3.capture2e(*cmd)
+          if status.exitstatus != 0
+            raise UncompilablePayloadError, stdout_err
+          end
+          stdout_err
+        end
 
-          unless self.keep_src
-            File.delete(src_file) if File.exist?(src_file)
+        def cleanup_files(files = [])
+          unless keep_src
+            src_file = "#{file_name}.c"
+            files << src_file
           end
 
-          unless self.keep_exe
-            File.delete(exe_file) if File.exist?(exe_file)
+          unless keep_exe
+            exe_file = "#{file_name}.exe"
+            files << exe_file
+          end
+
+          files.each do |file|
+            File.delete(file) if File.exist?(file)
```

Comment:
None of your code uses this new code atm, so I'm confused why the function signature has been changed and this code added, as it won't clean up the CPP files.

Reply:
it's exposed for future module implementors to use, if they so wish. 
you recommended i modify it to account for the new code, but there's nothing to clean since tempfiles and tempdirs manage themselves

Reply:
reverted